### PR TITLE
Allow training type plugin to delay optimizer creation (FSDP 2/n)

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Allows the training type plugin to delay optimizer creation. This is useful when the optimizers can only be created after the model has been wrapped. In the case of FSDP, after the model has been sharded onto devices.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 1]
```diff
@@ -169,3 +169,12 @@ def init_optimizers(self, trainer: "Trainer", model: LightningModule):
...
         optimizer.step(closure=lambda_closure, **kwargs)
+
+    @property
+    def setup_optimizers_after_dispatch(self) -> bool:
+        """
+        Override to delay setting optimizers and schedulers till after dispatch.
+        This is useful when the `TrainingTypePlugin` requires operating on the wrapped accelerator model.
+        Returns: True if delaying setup optimizers till after dispatch, False to call within setup.
+        """
+        return False
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:180]: do we want such property we even moved from it with model optimization...

File: tests/accelerators/test_cpu.py [hunk 2]
```diff
@@ -1,12 +1,13 @@
 from unittest.mock import Mock
 
 import pytest
+import pytorch_lightning as pl
```
Comment (Borda) [inline tests/accelerators/test_cpu.py:4]: import only needed

File: pytorch_lightning/plugins/training_type/training_type_plugin.py (hunk 1 above)
Comment (SeanNaren) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:180]: I'm unsure what you mean here

Comment (Borda) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:180]: I am quite sure you shall not be here :D go away from the computer...

Comment (Borda) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:180]: ok, it seems I was wrong and we still have it
https://github.com/PyTorchLightning/pytorch-lightning/blob/2708c3993dfc66989884d8e2971eba7036fcbd82/pytorch_lightning/core/lightning.py#L166

File: pytorch_lightning/accelerators/accelerator.py [hunk 3]
```diff
@@ -74,7 +74,8 @@ def setup(self, trainer: 'Trainer', model: LightningModule) -> None:
             model: the model to train
         """
         self.connect_training_type_plugin(self.training_type_plugin, model)
-        self.setup_optimizers(trainer)
+        if not self.training_type_plugin.setup_optimizers_after_dispatch:
```
Comment (tchaton) [inline pytorch_lightning/accelerators/accelerator.py:77]: Hey @SeanNaren 
1. Won't it be better to move `self.setup_optimizers(trainer)` in dispatch directly or there will be a blocking part ?
2. If we stick with `setup_optimizers_after_dispatch`, it would be more clear to have `setup_optimizers_in_pre_dispatch` ?

Comment (SeanNaren) [reply pytorch_lightning/accelerators/accelerator.py:77]: I never thought about it honestly, that's a really good point lol. Let me just move it and see what happens

Comment (SeanNaren) [reply pytorch_lightning/accelerators/accelerator.py:77]: After #6506 I understand this piece of code a bit better.

Because Apex wraps optimizers, the optimizers have to be set before setting up the precision plugin.

So I'll update the doc strings making it clear that if you enable `setup_optimizers_in_pre_dispatch` that this may break APEX.

**Review by awaelchli — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 4]
```diff
@@ -182,3 +182,13 @@ def init_optimizers(self, trainer: "Trainer", model: LightningModule):
 
     def optimizer_step(self, optimizer: torch.optim.Optimizer, lambda_closure: Callable, **kwargs):
         optimizer.step(closure=lambda_closure, **kwargs)
+
+    @property
+    def setup_optimizers_in_pre_dispatch(self) -> bool:
+        """
+        Override to delay setting optimizers and schedulers till after dispatch.
+        This is useful when the `TrainingTypePlugin` requires operating on the wrapped accelerator model.
+        However this may break certain precision plugins such as APEX which require optimizers to be set.
```
Comment (ananthsub) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:191]: n00b question: why does this break APEX?

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/accelerators/accelerator.py (hunk 3 above)
Comment (ananthsub) [reply pytorch_lightning/accelerators/accelerator.py]: total n00b question: is it possible to delay setting up both the optimizers and precision plugin to pre-dispatch to make apex work?

File: pytorch_lightning/plugins/training_type/training_type_plugin.py (hunk 4 above)
Comment (SeanNaren) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:191]: APEX wraps the optimizer into an FP16_Optimizer (I've sure you've seen this in a few places). The optimizer needs to be defined for this to work :(

Annoying because this is the only reason for this overcomplication within the training type plugin.

**Review by tchaton — APPROVED** [review]
LGTM !
